src/gt4py/next/type_system_2/types.py

```diff
@@ -0,0 +1,257 @@
+from . import traits
+import dataclasses
+from typing import Optional, Any, Sequence
+
+from .traits import FunctionArgument
+
+
+class Type:
+    ...
+
+
+@dataclasses.dataclass
+class IntegerType(
+    Type,
+    traits.SignednessTrait,
+    traits.FromTrait,
+    traits.FromImplicitTrait,
+    traits.ArithmeticTrait,
+    traits.BitwiseTrait
+):
+    width: int
+    signed: bool
+
+    def __init__(self, width: int, signed: bool):
+        assert (width in [1, 8, 16, 32, 64])
+        if width == 1:
+            assert (signed == False)
+        self.width = width
+        self.signed = signed
+
+    def __str__(self):
+        signed = "" if self.signed else "u"
+        return f"{signed}int{self.width}" if self.width > 1 else "bool"
+
+    def is_signed(self) -> bool:
+        return self.signed
+
+    def is_constructible_from(self, ty: Type) -> bool:
+        return isinstance(ty, IntegerType) or isinstance(ty, FloatType)
+
+    def is_implicitly_constructible_from(self, ty: Any) -> bool:
+        if isinstance(ty, IntegerType):
+            if self.signed:
+                return (ty.signed and ty.width <= self.width) or (not ty.signed and ty.width < self.width)
+            else:
+                return not ty.signed and ty.width <= self.width
+        return False
+
+    def common_arithmetic_type(self, other: Type) -> Optional[Type]:
+        return self.common_bitwise_type(other)
+
+    def common_bitwise_type(self, other: Type) -> Optional[Type]:
+        if isinstance(other, IntegerType):
+            if not self.signed and other.signed:
+                return None
+            elif self.signed and not other.signed:
+                required_width = max(self.width, 2 * other.width)
+                return IntegerType(required_width, True) if required_width <= 64 else None
+            else:
+                required_width = max(self.width, other.width)
+                return IntegerType(required_width, self.signed)
+        return None
+
+
+@dataclasses.dataclass
+class FloatType(
+    Type,
+    traits.SignednessTrait,
+    traits.FromTrait,
+    traits.FromImplicitTrait,
+    traits.ArithmeticTrait
+):
+    width: int
+
+    def __init__(self, width: int):
+        assert (width in [16, 32, 64])
+        self.width = width
+
+    def __str__(self):
+        return f"float{self.width}"
+
+    def is_signed(self) -> bool:
+        return True
+
+    def is_constructible_from(self, ty: Type) -> bool:
+        return isinstance(ty, IntegerType) or isinstance(ty, FloatType)
+
+    def is_implicitly_constructible_from(self, ty: Any) -> bool:
+        if isinstance(ty, IntegerType):
+            return ty.width < self.width
+        elif isinstance(ty, FloatType):
+            return ty.width <= self.width
```

Comment:
At least for now, we want to be explicit of the floating point width, otherwise users might very easily cast to a larger floating point type without noticing.

Reply:
The purpose of this function is to express if a type can be losslessly converted to another. The IR does not have to use this, it can also use equality comparison if we want to disallow lossless implicit type conversions as well.